merge wip/master

schmir opened the pull request:
I'd say this can be used on the kovan testnet. Please have a look at it.

We can do regular code reviews from now on.

schmir committed:
import the first real code

ethidx.logdecode contains code to decode events from eth logs, we we get with
web3.eth.getLogs for a set of contract addresses.

ethidx.pgimport contains a first version that imports those events into a table
containing a jsonb column holding the events parameters.

this needs more work, but we're already able to import events into postgres.

schmir committed:
add pgimport.insert_events function

schmir committed:
add short demo program showing a query on a jsonb field

Eventually will end up in the documentation.

schmir committed:
enrich events with information from blocks

we need the timestamp for each event, which is only stored in the block.

So, after fetching events, we fetch each block by block number that contains an
event.

Strictly speaking we could have one or multiple chain reorgs while fetching
those blocks, so we compare the block hash with the one stored in the event and
bail out if they don't match.

I had already been introducing a new table 'blocks', which would hold some basic
information about blocks. But then decided against it, cause it's a bit easier
to implement this way, i.e. with a single table.

Also, we now use psycopg2.extras.RealDictCursor. This makes psycopg2 give us
results as a dictionary (instead of a tuple).

schmir committed:
use RealDictCursor in run-query

we want the results to be returned as dict.

schmir committed:
define a primary key for events

schmir committed:
rename ethidx to ethindex

as requested by bernd.

schmir committed:
store ABIs in database

add an ethindex program. 'ethindex importabi' will imports ABI as used by the
relay server into the database.

schmir committed:
implement incremental imports

This change implements a first version with incremental indexing. I have to
clean it up and add some documentation.

schmir committed:
add tox configuration

run flake8, black and mypy via tox. This already found a bug.
pycryptodome 3.6.2 vanished from PyPI. So, upgrade that to 3.6.3

schmir committed:
rename the package to py-eth-index, fix project URL

schmir committed:
run black on travis

run black on travis. We now also try to install the non-binary psycopg2.
black is also being added to requirements.txt, because pyflakes and mypy are
also there.

schmir committed:
add a Dockerfile

This Dockerfile builds the currently checked out version.

schmir committed:
add --jsonrpc option

schmir committed:
show version of eth-index when running programs

schmir committed:
do not get events for the latest block again

The upper bound of the getLogs call is inclusive. Do not get events for that
last block again.

The log message may get a bit confusing now with the fromBlock being greater
than the toBlock parameter, like in:

,----
| INFO:ethindex.pgimport:got 0 events in 0 blocks (7719357 -> 7719356)
`----

schmir committed:
upgrade constraints.txt to the latest versions available

this makes ethindex work with python 3.7.

schmir committed:
refactor actual syncing into _sync_blocks method

and log a message if we're already synced

schmir committed:
install flake8 plugins that we also use in the relay server

schmir committed:
describe installation and usage

schmir committed:
make runsync more robust

we catch errors in the outer code and are able to handle temporary outages of
postgresql and geth/parity that way.

The disadvantage is that it's now also more robust against configuration errors,
i.e.. it may loop forever if the configuration is wrong, e.g. by setting an
illegal hostname in PGHOST.

schmir committed:
add --waittime option to runsync

we may need this for the tests

schmir committed:
show command line arguments in readme

schmir committed:
add config file for pre-commit

pre-commit [1] is framework for managing git pre-commit hooks. Since I failed to
reformat files with black before committing them in the past, let's try
this. After installation of pre-commit, this can be activated by running

  pre-commit install

also fix some things that it already complained about.

[1] https://pre-commit.com/

berndbohmeier commented on ethindex/logdecode.py:
```diff
@@ -0,0 +1,88 @@
+import eth_utils
+
+
+def build_address_to_abi_dict(addresses_json, compiled_contracts):
```
Would be a good fit for type hints.

berndbohmeier commented on ethindex/logdecode.py:
```diff
@@ -0,0 +1,88 @@
+    """
+    res = {}
+
+    def add_abi(a, n):
```
a, n? constract_address, contract_name? Little more descriptive names should be possible.

berndbohmeier commented on ethindex/logdecode.py:
```diff
@@ -0,0 +1,88 @@
+    """decode non-indexed inputs from a log entry
+
+    This one decodes the values stored in the 'data' field."""
+    inputs = [x for x in abi["inputs"] if not x["indexed"]]
```
more descriptive names pls
e.g.
`non_indexed_inputs = [input_ for input_ in abi["inputs"] if not input_["indexed"]]`

berndbohmeier commented on ethindex/logdecode.py:
```diff
@@ -0,0 +1,88 @@
+    return zip(names, values)
+
+
+def filter_events(abi):
```
better function name. From filter_event I would expect a function that filters events, but this filters the abi for eventabis.
So maybe `get_eventabis_from_abi`? Or `get_event_abis`

berndbohmeier commented on ethindex/logdecode.py:
```diff
@@ -0,0 +1,88 @@
+        self.addresses = list(address2abi.keys())
+        self.address2abi = address2abi
+        self.address_topic2event_abi = {}
+        for a, abi in self.address2abi.items():
```
a -> address

berndbohmeier commented on ethindex/logdecode.py:
```diff
@@ -0,0 +1,88 @@
+    def decode_log(self, log):
+        abi = self.get_abi_for_log(log)
+        if abi is None:
+            return None
```
I think we should better throw here, or is it not a error if we can not find the abi?

schmir replied:
probably. This looks like it may be related to the bug you've seen with the clientlib tests...

berndbohmeier commented on ethindex/logdecode.py:
```diff
@@ -0,0 +1,88 @@
+            )
+        )
+
+        return {"name": abi["name"], "args": args, "log": log}
```
Can we maybe get a event class, at least something like
`Event = namedtuple('Event', 'name, args, blockNumber, logIndex... usw)`
would help the typechecker also in insert events

berndbohmeier commented on ethindex/pgimport.py:
```diff
@@ -0,0 +1,266 @@
+            )
+
+
+def event_blocknumbers(events):
```
type hints

berndbohmeier commented on ethindex/pgimport.py:
```diff
@@ -0,0 +1,266 @@
+    return psycopg2.connect(dsn, cursor_factory=psycopg2.extras.RealDictCursor)
+
+
+def enrich_events(events, blocks):
```
short doc string

berndbohmeier commented on ethindex/pgimport.py:
```diff
@@ -0,0 +1,266 @@
+        e["timestamp"] = block["timestamp"]
+
+
+def main():
```
Remove if not used

schmir committed:
make sure we store checksum addresses

we like to store checksum addresses for every value of type 'address', so let's
convert them after decoding.

schmir committed:
remove unused pgimport.main method

schmir committed:
remove unused pgimport.insert_blocks function

schmir committed:
add logdecode.Event class

TopicIndex.decode_log now returns instances of that class instead of dict.

I'm using the attrs module for that class. namedtuple wouldn't allow me to
define read-only properties that reach into the 'log' dictionary.

berndbohmeier committed:
Add workdir to know where files are searched for

berndbohmeier committed:
Add command to create tables

berndbohmeier committed:
Remove broken flake8 plugin

berndbohmeier committed:
Raise Exception when log abi not found

berndbohmeier committed:
Python >=3.6 is requiered

berndbohmeier committed:
Refactored code

Add some type hints
Better variable names

berndbohmeier committed:
Black formatting fixes
